### ISSUE TITLE
Automatic switch between AP and client?

### ISSUE THREAD

**Issue by ekawahyu** [mozilla-iot/gateway#1899]
I am running the gateway battery powered, I did all the WiFi setup to get it connected to my office network. When I move out of my office WiFi coverage, the gateway does not switch automatically to AP mode. Unless, of course, the AP mode will show up after power cycling it. Questions:
1. Is this intended for any reason?
2. Am I missing any setup/configuration somewhere?
3. How can I make automatic switching to work when the gateway is in/out of the WiFi coverage?

**Comment by benfrancis** [comment]
I actually thought it did. Though I've sometimes wondered about the security implications of this.

@mrstegeman Would you have expected this to work?

**Comment by mrstegeman** [comment]
It should revert to AP mode after a power cycle, if not connected to Ethernet.

**Comment by ekawahyu** [comment]
@mrstegeman Yes, it enters into AP mode only after a power cycle. But in my scenario, the gateway is constantly powered, it never turns off or reboot. And there is no ethernet connection at all time.

The gateway is mobile, battery powered, so it is constantly powered. When the gateway is moving from one building to another and these two buildings have different SSID and password, the gateway needs, at least, to switch into AP mode automatically for reconfiguration. But this is also an issue with 0.8.1, that it only allows one WiFi connection, not multiple.

You may ask what kind of scenario that requires the gateway to be constantly powered with multiple WiFi configured in it? I have a truck that loads/unloads at multiple locations. The gateway is in the individual truck that is made of Raspberry Pi and there are several sensors in the truck that constantly sending data to this gateway. When the truck is parked at location A, it gets connected to the (known) WiFi router in that location. All of the recorded data during transit gets uploaded to a server. The next time it parks at different location and there is (known) WiFi in that location, it gets connected to it and do the same thing over and over again.

My question is... would this kind of application fit to use WebThings Gateway?

**Comment by benfrancis** [comment]
@ekawahyu To be honest this is not an intended use case for Mozilla's gateway implementation, which is focused on the smart home. In the smart home environment it is assumed that the gateway is usually mains-powered and connected to a single Wi-Fi network (and in the case of router builds, *is* the Wi-Fi network).

Having the gateway switch between multiple saved Wi-Fi networks automatically would probably require complicating the network settings UI. We do have a similar use case when moving the gateway between Wi-Fi networks at conferences, but that always involves a power cycle.

I'm not sure what the current rationale is for only switching to AP mode on reboot?

This has also got me wondering about the risk of someone hijacking the gateway by connecting it to their own network while it's in AP mode. The rest of the gateway UI would still be password protected so this is more of an inconvenience than a security risk, but currently would require quite a manual process for the owner to fix.

**Comment by dhylands** [comment]
I think that the Raspberry Pi should switch Wifi networks automatically if the entries exist in the wpa_supplicant file, but that really needs to be tested.

**Comment by ekawahyu** [comment]
@benfrancis Noted! Thank you for confirming it. You may close this issue if you wish. I also noted that there is another issue if user clicks on the 'Skip' ... the gateway may never enter the AP mode ever again? But that would be another issue thread to submit. Thank you.

**Comment by ekawahyu** [comment]
@dhylands When the wpa_supplicant is edited by hand, yes, it does. But, it is gone once you use the WebThings UI to connect it to a different AP.

**Comment by benfrancis** [comment]
@dhylands Oh, cool. That would be a useful workaround.

@ekawahyu Can you connect the gateway to one network, then manually edit wpa_supplicant to add a second?

**Comment by ekawahyu** [comment]
@benfrancis Oh yeah, that's what I have been doing with all of my setup. The WebThings gateway makes it easier because I don't have to do headless setup anymore. That's what I like about it.
